File: tools/llm_client.py
"""
LLM backends for Rogue Rivals LLM agents (Anthropic, OpenAI, Z.AI, Groq, OpenRouter).
"""

from __future__ import annotations

import json
import os
import re
from pathlib import Path
from typing import Any, Dict, Optional

try:
    import jsonschema
except ImportError:  # pragma: no cover
    jsonschema = None  # type: ignore


def _load_dotenv_if_present() -> None:
    """Load KEY=VALUE pairs from repo-root .env into os.environ if not already set.

    Lightweight replacement for python-dotenv; avoids adding a dep.
    Silent no-op if .env is missing or unreadable.
    """
    env_path = Path(__file__).resolve().parent.parent / ".env"
    if not env_path.is_file():
        return
    try:
        for raw in env_path.read_text(encoding="utf-8").splitlines():
            line = raw.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, _, val = line.partition("=")
            key = key.strip()
            val = val.strip().strip("'").strip('"')
            if key and key not in os.environ:
                os.environ[key] = val
    except OSError:
        return


_load_dotenv_if_present()


class LLMError(Exception):
    """API failure or unrecoverable parse error."""


def _strip_json_fence(text: str) -> str:
    t = text.strip()
    m = re.match(r"^```(?:json)?\s*([\s\S]*?)```\s*$", t, re.IGNORECASE)
    if m:
        return m.group(1).strip()
    return t


def _validate_schema(data: Any, schema: Optional[dict]) -> None:
    if schema is None or jsonschema is None:
        return
    jsonschema.validate(instance=data, schema=schema)


ZAI_BASE_URL = "https://api.z.ai/api/paas/v4/"
ZAI_DEFAULT_MODEL = "glm-4.5-air"

GROQ_BASE_URL = "https://api.groq.com/openai/v1"
GROQ_DEFAULT_MODEL = "llama-3.3-70b-versatile"

OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"
OPENROUTER_DEFAULT_MODEL = "nvidia/nemotron-3-super-120b-a12b:free"


def _groq_key() -> str:
    return os.environ.get("GROQ_API_KEY", "").strip()


def _openrouter_key() -> str:
    return os.environ.get("OPENROUTER_API_KEY", "").strip()


def _zai_key() -> str:
    """Z.AI key may be provided under either ZAI_API_KEY or ZAI_KEY."""
    return (
        os.environ.get("ZAI_API_KEY", "").strip()
        or os.environ.get("ZAI_KEY", "").strip()
    )


def _pick_provider_model(
    provider: Optional[str], model: Optional[str]
) -> tuple[str, str, bool]:
    """Returns (provider, model, used_anthropic_preference).

    Provider selection priority when no explicit `provider` arg is passed:
      1. `LLM_PROVIDER` env var (anthropic | openai | zai | groq | openrouter)
      2. First available key in this order: anthropic, zai, openai, openrouter, groq
    """
    has_a = bool(os.environ.get("ANTHROPIC_API_KEY", "").strip())
    has_o = bool(os.environ.get("OPENAI_API_KEY", "").strip())
    has_z = bool(_zai_key())
    has_g = bool(_groq_key())
    has_or = bool(_openrouter_key())
    if not (has_a or has_o or has_z or has_g or has_or):
        raise LLMError(
            "No LLM API key configured. Set ANTHROPIC_API_KEY, OPENAI_API_KEY, "
            "ZAI_API_KEY (alias: ZAI_KEY), GROQ_API_KEY, or OPENROUTER_API_KEY."
        )

    effective = provider or os.environ.get("LLM_PROVIDER", "").strip() or None
    if effective:
        p = effective.lower()
        if p == "anthropic" and not has_a:
            raise LLMError("ANTHROPIC_API_KEY is not set.")
        if p == "openai" and not has_o:
            raise LLMError("OPENAI_API_KEY is not set.")
        if p == "zai" and not has_z:
            raise LLMError("ZAI_API_KEY / ZAI_KEY is not set.")
        if p == "groq" and not has_g:
            raise LLMError("GROQ_API_KEY is not set.")
        if p == "openrouter" and not has_or:
            raise LLMError("OPENROUTER_API_KEY is not set.")
        if p not in ("anthropic", "openai", "zai", "groq", "openrouter"):
            raise LLMError(f"Unknown provider: {effective}")
        if p == "anthropic":
            m = model or os.environ.get(
                "ANTHROPIC_MODEL", "claude-3-5-haiku-20241022"
            )
        elif p == "zai":
            m = model or os.environ.get("ZAI_MODEL", ZAI_DEFAULT_MODEL)
        elif p == "groq":
            m = model or os.environ.get("GROQ_MODEL", GROQ_DEFAULT_MODEL)
        elif p == "openrouter":
            m = model or os.environ.get("OPENROUTER_MODEL", OPENROUTER_DEFAULT_MODEL)
        else:
            m = model or os.environ.get("OPENAI_MODEL", "gpt-4o-mini")
        return p, m, p == "anthropic"

    if has_a:
        return (
            "anthropic",
            model or os.environ.get("ANTHROPIC_MODEL", "claude-3-5-haiku-20241022"),
            True,
        )
    if has_z:
        return (
            "zai",
            model or os.environ.get("ZAI_MODEL", ZAI_DEFAULT_MODEL),
            False,
        )
    if has_o:
        return (
            "openai",
            model or os.environ.get("OPENAI_MODEL", "gpt-4o-mini"),
            False,
        )
    if has_or:
        return (
            "openrouter",
            model or os.environ.get("OPENROUTER_MODEL", OPENROUTER_DEFAULT_MODEL),
            False,
        )
    return (
        "groq",
        model or os.environ.get("GROQ_MODEL", GROQ_DEFAULT_MODEL),
        False,
    )


class LLMClient:
    """LLM client (Anthropic, OpenAI-compatible, Z.AI, Groq, OpenRouter); returns JSON dict."""

    def __init__(
        self,
        provider: str = "",
        model: str = "",
        temperature: float = 0.0,
        max_input_tokens: int = 2000,
        max_output_tokens: int = 500,
    ):
        self.provider, self.model, _ = _pick_provider_model(
            provider or None, model or None
        )
        self.temperature = temperature
        self.max_input_tokens = max_input_tokens
        self.max_output_tokens = max_output_tokens

        if self.provider == "anthropic":
            try:
                from anthropic import Anthropic  # type: ignore
            except ImportError as e:
                raise LLMError("Install anthropic package: pip install anthropic") from e
            self._anthropic = Anthropic()
            self._openai = None
        elif self.provider == "zai":
            try:
                from openai import OpenAI  # type: ignore
            except ImportError as e:
                raise LLMError("Install openai package: pip install openai") from e
            zai_key = _zai_key()
            if not zai_key:
                raise LLMError("ZAI_API_KEY / ZAI_KEY is not set.")
            self._openai = OpenAI(api_key=zai_key, base_url=ZAI_BASE_URL)
            self._anthropic = None
        elif self.provider == "groq":
            try:
                from openai import OpenAI  # type: ignore
            except ImportError as e:
                raise LLMError("Install openai package: pip install openai") from e
            gk = _groq_key()
            if not gk:
                raise LLMError("GROQ_API_KEY is not set.")
            self._openai = OpenAI(api_key=gk, base_url=GROQ_BASE_URL)
            self._anthropic = None
        elif self.provider == "openrouter":
            try:
                from openai import OpenAI  # type: ignore
            except ImportError as e:
                raise LLMError("Install openai package: pip install openai") from e
            rk = _openrouter_key()
            if not rk:
                raise LLMError("OPENROUTER_API_KEY is not set.")
            referer = os.environ.get(
                "OPENROUTER_HTTP_REFERER", "https://localhost"
            ).strip()
            title = os.environ.get("OPENROUTER_APP_TITLE", "Rogue Rivals").strip()
            self._openai = OpenAI(
                api_key=rk,
                base_url=OPENROUTER_BASE_URL,
                default_headers={
                    "HTTP-Referer": referer,
                    "X-Title": title,
                },
            )
            self._anthropic = None
        else:
            try:
                from openai import OpenAI  # type: ignore
            except ImportError as e:
                raise LLMError("Install openai package: pip install openai") from e
            self._openai = OpenAI()
            self._anthropic = None

    def _truncate_text(self, text: str, budget_tokens: int) -> str:
        # ~4 chars per token heuristic for Latin text
        max_chars = max(256, budget_tokens * 4)
        if len(text) <= max_chars:
            return text
        return text[: max_chars - 40] + "\n...[truncated for token cap]"

    def complete(
        self,
        system: str,
        user: str,
        schema: Optional[dict] = None,
    ) -> Dict[str, Any]:
        sys_b = self._truncate_text(system, self.max_input_tokens // 2)
        usr_budget = max(256, self.max_input_tokens - len(sys_b) // 4)
        usr_b = self._truncate_text(user, usr_budget)

        t0 = __import__("time").perf_counter()
        raw_text = ""
        in_tok = out_tok = 0

        try:
            if self.provider == "anthropic":
                msg = self._anthropic.messages.create(
                    model=self.model,
                    max_tokens=self.max_output_tokens,
                    temperature=self.temperature,
                    system=sys_b,
                    messages=[{"role": "user", "content": usr_b}],
                )
                raw_text = "".join(
                    b.text for b in msg.content if getattr(b, "type", "") == "text"
                )
                in_tok = getattr(msg.usage, "input_tokens", 0) or 0
                out_tok = getattr(msg.usage, "output_tokens", 0) or 0
            else:
                kwargs: Dict[str, Any] = dict(
                    model=self.model,
                    temperature=self.temperature,
                    max_tokens=self.max_output_tokens,
                    messages=[
                        {"role": "system", "content": sys_b},
                        {"role": "user", "content": usr_b},
                    ],
                )
                if self.provider in ("openai", "groq"):
                    kwargs["response_format"] = {"type": "json_object"}
                elif self.provider == "openrouter":
                    # Some OpenRouter models reject json_object; allow opt-out.
                    om = os.environ.get("OPENROUTER_JSON_MODE", "true").strip().lower()
                    if om not in ("0", "false", "no", "off"):
                        kwargs["response_format"] = {"type": "json_object"}
                elif self.provider == "zai":
                    # GLM-4.x reasoning models default to `thinking.enabled`,
                    # which consumes the output budget with hidden CoT. For
                    # low-latency action decisions we disable it unless the
                    # user explicitly opts back in via ZAI_THINKING=enabled.
                    zai_thinking = os.environ.get(
                        "ZAI_THINKING", "disabled"
                    ).strip().lower()
                    if zai_thinking in ("enabled", "disabled"):
                        kwargs["extra_body"] = {
                            "thinking": {"type": zai_thinking}
                        }
                comp = self._openai.chat.completions.create(**kwargs)
                raw_text = comp.choices[0].message.content or ""
                if comp.usage:
                    in_tok = comp.usage.prompt_tokens or 0
                    out_tok = comp.usage.completion_tokens or 0
        except Exception as e:
            raise LLMError(str(e)) from e

        latency_ms = int((__import__("time").perf_counter() - t0) * 1000)

        try:
            data = json.loads(_strip_json_fence(raw_text))
        except json.JSONDecodeError as e:
            raise LLMError(f"Invalid JSON from model: {e}") from e

        try:
            _validate_schema(data, schema)
        except Exception as e:
            raise LLMError(f"Schema validation failed: {e}") from e

        data["_usage"] = {
            "input_tokens": in_tok,
            "output_tokens": out_tok,
            "latency_ms": latency_ms,
            "model": self.model,
            "provider": self.provider,
        }
        return data

File: packages/llm/src/llmClient.ts
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";
import { stripJsonFence } from "./stripJsonFence.js";

export class LLMError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LLMError";
  }
}

const ZAI_BASE_URL = "https://api.z.ai/api/paas/v4/";
const ZAI_DEFAULT_MODEL = "glm-4.5-air";
const GROQ_BASE_URL = "https://api.groq.com/openai/v1";
const GROQ_DEFAULT_MODEL = "llama-3.3-70b-versatile";
const OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1";
const OPENROUTER_DEFAULT_MODEL = "nvidia/nemotron-3-super-120b-a12b:free";

function groqKey(): string {
  return (process.env.GROQ_API_KEY ?? "").trim();
}
function openrouterKey(): string {
  return (process.env.OPENROUTER_API_KEY ?? "").trim();
}
function zaiKey(): string {
  return (process.env.ZAI_API_KEY ?? process.env.ZAI_KEY ?? "").trim();
}

export type LlmProvider = "anthropic" | "openai" | "zai" | "groq" | "openrouter";

/** Loose validation aligned with ORDER_PACKET_SCHEMA (tools/v2/llm_agent.py). */
function validateOrderPacketLoose(data: Record<string, unknown>, _schema: object): void {
  if (data.choose !== undefined) {
    if (!Array.isArray(data.choose)) throw new LLMError("Schema validation failed: choose must be an array");
    if (data.choose.length > 12) throw new LLMError("Schema validation failed: choose maxItems 12");
    for (const x of data.choose) {
      if (typeof x !== "string") throw new LLMError("Schema validation failed: choose items must be strings");
    }
  }
  if (data.messages !== undefined) {
    if (!Array.isArray(data.messages)) throw new LLMError("Schema validation failed: messages must be an array");
    if (data.messages.length > 8) throw new LLMError("Schema validation failed: messages maxItems 8");
    for (const m of data.messages) {
      if (m === null || typeof m !== "object") throw new LLMError("Schema validation failed: message entry invalid");
      const o = m as Record<string, unknown>;
      if (typeof o.to !== "string" || typeof o.text !== "string") {
        throw new LLMError("Schema validation failed: messages items need to, text");
      }
    }
  }
  if (data.orders !== undefined) {
    if (!Array.isArray(data.orders)) throw new LLMError("Schema validation failed: orders must be an array");
    if (data.orders.length > 12) throw new LLMError("Schema validation failed: orders maxItems 12");
  }
}

export function pickProviderModel(
  provider: string | undefined,
  model: string | undefined,
): { provider: LlmProvider; model: string; anthropicPreference: boolean } {
  const hasA = Boolean((process.env.ANTHROPIC_API_KEY ?? "").trim());
  const hasO = Boolean((process.env.OPENAI_API_KEY ?? "").trim());
  const hasZ = Boolean(zaiKey());
  const hasG = Boolean(groqKey());
  const hasOr = Boolean(openrouterKey());
  if (!hasA && !hasO && !hasZ && !hasG && !hasOr) {
    throw new LLMError(
      "No LLM API key configured. Set ANTHROPIC_API_KEY, OPENAI_API_KEY, " +
        "ZAI_API_KEY (alias: ZAI_KEY), GROQ_API_KEY, or OPENROUTER_API_KEY.",
    );
  }

  const effective = (provider ?? process.env.LLM_PROVIDER ?? "").trim() || undefined;
  if (effective) {
    const p = effective.toLowerCase();
    if (p === "anthropic" && !hasA) throw new LLMError("ANTHROPIC_API_KEY is not set.");
    if (p === "openai" && !hasO) throw new LLMError("OPENAI_API_KEY is not set.");
    if (p === "zai" && !hasZ) throw new LLMError("ZAI_API_KEY / ZAI_KEY is not set.");
    if (p === "groq" && !hasG) throw new LLMError("GROQ_API_KEY is not set.");
    if (p === "openrouter" && !hasOr) throw new LLMError("OPENROUTER_API_KEY is not set.");
    if (!["anthropic", "openai", "zai", "groq", "openrouter"].includes(p)) {
      throw new LLMError(`Unknown provider: ${effective}`);
    }
    let m: string;
    if (p === "anthropic")
      m = model ?? process.env.ANTHROPIC_MODEL ?? "claude-3-5-haiku-20241022";
    else if (p === "zai") m = model ?? process.env.ZAI_MODEL ?? ZAI_DEFAULT_MODEL;
    else if (p === "groq") m = model ?? process.env.GROQ_MODEL ?? GROQ_DEFAULT_MODEL;
    else if (p === "openrouter") m = model ?? process.env.OPENROUTER_MODEL ?? OPENROUTER_DEFAULT_MODEL;
    else m = model ?? process.env.OPENAI_MODEL ?? "gpt-4o-mini";
    return { provider: p as LlmProvider, model: m, anthropicPreference: p === "anthropic" };
  }

  if (hasA) {
    return {
      provider: "anthropic",
      model: model ?? process.env.ANTHROPIC_MODEL ?? "claude-3-5-haiku-20241022",
      anthropicPreference: true,
    };
  }
  if (hasZ) {
    return { provider: "zai", model: model ?? process.env.ZAI_MODEL ?? ZAI_DEFAULT_MODEL, anthropicPreference: false };
  }
  if (hasO) {
    return { provider: "openai", model: model ?? process.env.OPENAI_MODEL ?? "gpt-4o-mini", anthropicPreference: false };
  }
  if (hasOr) {
    return {
      provider: "openrouter",
      model: model ?? process.env.OPENROUTER_MODEL ?? OPENROUTER_DEFAULT_MODEL,
      anthropicPreference: false,
    };
  }
  return {
    provider: "groq",
    model: model ?? process.env.GROQ_MODEL ?? GROQ_DEFAULT_MODEL,
    anthropicPreference: false,
  };
}

/**
 * Throws {@link LLMError} if the current `process.env` cannot supply a provider
 * for {@link LLMClient} (no API keys, or {@link LLM_PROVIDER} points at a missing key).
 * Use before starting LLM-backed matches so misconfiguration fails immediately.
 */
export function assertLlmEnvironmentConfigured(): void {
  pickProviderModel(undefined, undefined);
}

export interface LlmUsageMeta {
  input_tokens: number;
  output_tokens: number;
  latency_ms: number;
  model: string;
  provider: string;
}

export interface LlmCompleteResult {
  data: Record<string, unknown>;
  rawResponse: string;
  usage: LlmUsageMeta;
}

export class LLMClient {
  readonly provider: LlmProvider;
  readonly model: string;
  readonly temperature: number;
  readonly maxInputTokens: number;
  readonly maxOutputTokens: number;
  private anthropic: Anthropic | null = null;
  private openai: OpenAI | null = null;

  constructor(opts?: {
    provider?: string;
    model?: string;
    temperature?: number;
    maxInputTokens?: number;
    maxOutputTokens?: number;
  }) {
    const picked = pickProviderModel(opts?.provider, opts?.model);
    this.provider = picked.provider;
    this.model = picked.model;
    this.temperature = opts?.temperature ?? 0;
    this.maxInputTokens = opts?.maxInputTokens ?? 4000;
    this.maxOutputTokens = opts?.maxOutputTokens ?? 700;

    if (this.provider === "anthropic") {
      this.anthropic = new Anthropic();
    } else if (this.provider === "zai") {
      const k = zaiKey();
      if (!k) throw new LLMError("ZAI_API_KEY / ZAI_KEY is not set.");
      this.openai = new OpenAI({ apiKey: k, baseURL: ZAI_BASE_URL });
    } else if (this.provider === "groq") {
      const gk = groqKey();
      if (!gk) throw new LLMError("GROQ_API_KEY is not set.");
      this.openai = new OpenAI({ apiKey: gk, baseURL: GROQ_BASE_URL });
    } else if (this.provider === "openrouter") {
      const rk = openrouterKey();
      if (!rk) throw new LLMError("OPENROUTER_API_KEY is not set.");
      const referer = (process.env.OPENROUTER_HTTP_REFERER ?? "https://localhost").trim();
      const title = (process.env.OPENROUTER_APP_TITLE ?? "Rogue Rivals").trim();
      this.openai = new OpenAI({
        apiKey: rk,
        baseURL: OPENROUTER_BASE_URL,
        defaultHeaders: { "HTTP-Referer": referer, "X-Title": title },
      });
    } else {
      this.openai = new OpenAI();
    }
  }

  private truncateText(text: string, budgetTokens: number): string {
    const maxChars = Math.max(256, budgetTokens * 4);
    if (text.length <= maxChars) return text;
    return `${text.slice(0, maxChars - 40)}\n...[truncated for token cap]`;
  }

  async complete(
    system: string,
    user: string,
    schema: object | undefined,
  ): Promise<LlmCompleteResult> {
    const sysB = this.truncateText(system, this.maxInputTokens >> 1);
    const usrBudget = Math.max(256, this.maxInputTokens - Math.floor(sysB.length / 4));
    const usrB = this.truncateText(user, usrBudget);

    const t0 = performance.now();
    let rawText = "";
    let inTok = 0;
    let outTok = 0;

    try {
      if (this.provider === "anthropic" && this.anthropic) {
        const msg = await this.anthropic.messages.create({
          model: this.model,
          max_tokens: this.maxOutputTokens,
          temperature: this.temperature,
          system: sysB,
          messages: [{ role: "user", content: usrB }],
        });
        rawText = msg.content
          .filter((b) => b.type === "text")
          .map((b) => ("text" in b ? b.text : ""))
          .join("");
        inTok = msg.usage?.input_tokens ?? 0;
        outTok = msg.usage?.output_tokens ?? 0;
      } else if (this.openai) {
        const kwargs: OpenAI.Chat.ChatCompletionCreateParams = {
          model: this.model,
          temperature: this.temperature,
          max_tokens: this.maxOutputTokens,
          messages: [
            { role: "system", content: sysB },
            { role: "user", content: usrB },
          ],
        };
        if (this.provider === "openai" || this.provider === "groq") {
          kwargs.response_format = { type: "json_object" };
        } else if (this.provider === "openrouter") {
          const om = (process.env.OPENROUTER_JSON_MODE ?? "true").trim().toLowerCase();
          if (!["0", "false", "no", "off"].includes(om)) {
            kwargs.response_format = { type: "json_object" };
          }
        } else if (this.provider === "zai") {
          const zaiThinking = (process.env.ZAI_THINKING ?? "disabled").trim().toLowerCase();
          if (zaiThinking === "enabled" || zaiThinking === "disabled") {
            (kwargs as unknown as { extra_body?: object }).extra_body = {
              thinking: { type: zaiThinking },
            };
          }
        }
        const comp = await this.openai.chat.completions.create(kwargs);
        rawText = comp.choices[0]?.message?.content ?? "";
        if (comp.usage) {
          inTok = comp.usage.prompt_tokens ?? 0;
          outTok = comp.usage.completion_tokens ?? 0;
        }
      } else {
        throw new LLMError("LLM client not initialized");
      }
    } catch (e) {
      if (e instanceof LLMError) throw e;
      throw new LLMError(e instanceof Error ? e.message : String(e));
    }

    const latencyMs = Math.round(performance.now() - t0);

    let data: Record<string, unknown>;
    try {
      data = JSON.parse(stripJsonFence(rawText)) as Record<string, unknown>;
    } catch (e) {
      throw new LLMError(`Invalid JSON from model: ${e instanceof Error ? e.message : String(e)}`);
    }

    if (schema) {
      validateOrderPacketLoose(data, schema);
    }

    const usage = {
      input_tokens: inTok,
      output_tokens: outTok,
      latency_ms: latencyMs,
      model: this.model,
      provider: this.provider,
    } satisfies LlmUsageMeta;

    data._usage = usage;
    return { data, rawResponse: rawText, usage };
  }
}
